Mathlib/Topology/UniformSpace/Ultra.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2017 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Topology.UniformSpace.Defs
+import Mathlib.Topology.Bases
+
+/-!
+# Ultrametric (nonarchimedean) uniform spaces
+
+Ultrametric (nonarchimedean) uniform spaces are ones that generalize ultrametric spaces by
+having a uniformity based on equivalence relations.
+
+## Main definitions
+
+In this file we define `IsUltraUniformity`, a Prop mixin typeclass.
+
+## Main results
+
+* `TopologicalSpace.isTopologicalBasis_clopens`: a uniform space with a nonarchimedean uniformity
+  has a topological basis of clopen sets in the topology, meaning that it is topologically
+  zero-dimensional.
+
+## Implementation notes
+
+As in the `Topology/UniformSpace/Defs.lean` file, we do not reuse `Data/Rel.lean`
+but rather extend the relation properties as needed.
+
+## TODOs
+
+* Prove that `IsUltraUniformity` iff `IsUltrametricDist` on a `PseudoMetricSpace`
+* Generalize `IsUltrametricDist` to `IsUltrametricUniformity`
+* Provide `IsUltraUniformity` for the uniformity in a `Valued` ring
+* Generalize results about open/closed balls and spheres in `IsUltraUniformity` to
+  combine applications for `MetricSpace.ball` and valued "balls"
+* Use `IsUltraUniformity` to work with profinite/totally separated spaces
+* Define the nonarchimedean uniformity of a space that is a product of `IsUltraUniformity`s
+* Show that the `UniformSpace.Completion` of an `IsUltraUniformity` is `IsUltraUniformity`
+
+## References
+
+* [D. Windisch, *Equivalent characterizations of non-Archimedean uniform spaces*][windisch2021]
+* [A. C. M. van Rooij, *Non-Archimedean uniformities*][vanrooij1970]
+
+-/
+
+open Set Filter Topology
+open scoped Uniformity
+
+variable {X : Type*}
+
+/-- The relation is transitive. -/
+def IsTransitiveRel (V : Set (X × X)) : Prop :=
+  ∀ ⦃x y z⦄, (x, y) ∈ V → (y, z) ∈ V → (x, z) ∈ V
+
+lemma IsTransitiveRel.comp_eq_of_idRel_subset {s : Set (X × X)}
+    (h : IsTransitiveRel s) (h' : idRel ⊆ s) :
+    s ○ s = s := by
+  refine le_antisymm ?_ (subset_comp_self h')
+  intro ⟨x, y⟩
+  simp only [mem_compRel, forall_exists_index, and_imp]
+  intro z
+  exact @h x z y
+
+variable [UniformSpace X]
+
+variable (X) in
+/-- A uniform space is ultrametric if the uniformity `𝓤 X` has a basis of equivalence relations. -/
+class IsUltraUniformity : Prop where
+  has_basis : (𝓤 X).HasBasis
+    (fun s : Set (X × X) => s ∈ 𝓤 X ∧ IsSymmetricRel s ∧ IsTransitiveRel s) id
```

Comment:
Could you provide a constructor taking any basis made of equivalence relations?

Reply:
What kind of basis should we take? IsBasis, FilterBasis, IsTopologicalBasis, or something else? That this basis is used via UniformSpace.Core, or some other connection to the ambient uniform structure?

Reply:
Or just a rephrase that uses HasBasis with the additional reflexivity condition on sets?

Reply:
What I meant is, instead of considering the family of *all* entourages which are transitive and symmetric, assume that there is *some* basis for the uniformity filter (formulated using `HasBasis`) whose elements are transitive and symmetric.

Reply:
Let me just write down the statement I had in mind that will be clearer.

Reply:
Here it is:

```lean
lemma IsUltraUniformity.mk_of_hasBasis {ι : Type*} {p : ι → Prop} {s : ι → Set (X × X)}
    (h_basis : (𝓤 X).HasBasis p s) (h_symm : ∀ i, p i → IsSymmetricRel (s i))
    (h_trans : ∀ i, p i → IsTransitiveRel (s i)) :
    IsUltraUniformity X where
  has_basis := h_basis.to_hasBasis'
    (fun i hi ↦ ⟨s i, ⟨h_basis.mem_of_mem hi, h_symm i hi, h_trans i hi⟩, subset_rfl⟩)
    (fun _ hs ↦ hs.1)
```

The goal is simply that, if you happen to have your favorite basis for the uniformity (e.g nice subrings) which satisfy the conditions, it spares you from studying *all* the entourages satisfying the condition.